src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js

```diff
@@ -0,0 +1,711 @@
+    showAll: (e) => (e.ctrlKey || e.metaKey) && e.key === 'a',
+}
+
+const LABEL_STYLE = 'font-family: DejaVuSansMonoBook; font-size: 10px;'
+const MARGIN = 25
+const X_AXIS_LABEL_WIDTH = 10
+const Y_AXIS_LABEL_WIDTH = 10
+const ANIMATION_DURATION = 1000
+const LINEAR_SCALE = 'linear'
+const LIGHT_PLOT_COLOR = '#00E890'
+const DARK_PLOT_COLOR = '#E0A63B'
+const DEFAULT_NUMBER_OF_BINS = 10
+const BUTTON_HEIGHT = 25
+
+/**
+ * A d3.js histogram visualization.
+ *
+ *
+ * Data format (json):
+ {
+   "axis" : {
+      "x" : { "label" : "x-axis label", "scale" : "linear" },
+      "y" : { "label" : "y-axis label", "scale" : "logarithmic" },
+   },
+   "focus" { "x" : 1.7, "y" : 2.1, "zoom" : 3.0 },
+   "color" :  "rgb(1.0,0.0,0.0)" },
+   "bins" : 10,
+   "data" : [
+     "values" : [0.1, 0.2, 0.1, 0.15, 0.7],
+   ]
+ }
+ */
+class Histogram extends Visualization {
+    static inputType = 'Any'
+    static label = 'Histogram'
+
+    onDataReceived(data) {
+        const parsedData = JSON.parse(data)
+        const isUpdate = parsedData.update === 'diff'
+        this.updateState(parsedData, isUpdate)
+
+        if (!this.isInitialised()) {
+            this.initCanvas()
+            this.initLabels()
+            this.initHistogram()
+        }
+        this.updateLabels()
+        this.updateHistogram()
+    }
+
+    /**
+     * Indicates whether this visualisation has been initialised.
+     */
+    isInitialised() {
+        ok(this.svg)
+    }
+
+    /**
+     * Update the internal data and plot settings with the ones from the new incoming data.
+     * If no new settings/data have been provided the old ones will be kept.
+     */
+    updateState(data, isUpdate) {
+        if (isUpdate) {
+            this._axisSpec = ok(data.axis) ? data.axis : this._axisSpec
+            this._focus = ok(data.focus) ? data.focus : this._focus
+            this._dataValues = ok(data.data.values)
+                ? data.data.values
+                : this.data
+            this._bins = ok(data.bins) ? data.bins : this._bins
+        } else {
+            this._axisSpec = data.axis
+            this._focus = data.focus
+            this._dataValues = data.data.values
+            this._bins = data.bins
+        }
+    }
+
+    /**
+     * Return the focus area of the histogram. If none is set, returns undefined.
+     */
+    focus() {
+        return this._focus
+    }
+
+    /**
+     * Return specification for the axis. This includes scales (logarithmic/linear) and labels.
+     */
+    axisSpec() {
+        return (
+            this._axisSpec || {
+                x: { scale: LINEAR_SCALE },
+                y: { scale: LINEAR_SCALE },
+            }
+        )
+    }
+
+    /**
+     * Return vales to plot.
+     */
+    data() {
+        return this._dataValues || {}
+    }
+
+    /**
+     * Return the number of bins to use for the histogram..
+     */
+    binCount() {
+        return this._bins || DEFAULT_NUMBER_OF_BINS
+    }
```

Comment:
are you sure that if user sets 0 bins we should defautl to the default number of bins ?